### PR TITLE
ast: save origin as a private field, add `apply()`

### PR DESCRIPTION
**Pull request by kitbellew**
This is one of key changes enabling storing of origin for quasiquotes. Helps with #3425.

### CODE REVIEW

File: tests/shared/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala [hunk 1]
```diff
@@ -1121,11 +1120,21 @@ class SyntacticSuite extends scala.meta.tests.parsers.ParseSuite {
 
   test("interpolator braces for operator identifiers") {
     implicit def parseStat(code: String, dialect: Dialect): Stat = super.templStat(code)(dialect)
-    checkWithOriginalSyntax[Stat](q"""s"$${+++}bar"""")("""s"${+++}bar"""")
-    checkWithOriginalSyntax[Stat](q"""s"$${+++}_bar"""")("""s"${+++}_bar"""")
-    checkWithOriginalSyntax[Stat](q"""s"$${+++}123"""")("""s"${+++}123"""")
-    checkWithOriginalSyntax[Stat](q"""s"$${+++}***"""")("""s"${+++}***"""")
-    checkWithOriginalSyntax[Stat](q"""s"$${+++} ***"""")("""s"${+++} ***"""")
+    interceptMessage[NoSuchElementException]("token 11 out of 6") {
```
Comment (kitbellew) [inline tests/shared/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala:1123]: this reveals a bug which will be fixed in a separate commit.

**Comment by kitbellew** [comment]
@tgodzik this is the final commit which persists origin in quasiquotes, it marks origin with the new annotation (private field) and adds companion apply methods which take origin as extra parameter.

previous prs (notably, #3452 which includes fields marked with `@privateField` along with public ones and calls the expanded apply, and #3450 which enables handling of origin/input/dialect) made this possible.

**Review by tgodzik — APPROVED** [review]
LGTM
